Python/optimizer_analysis.c

```diff
@@ -496,6 +530,99 @@ peephole_opt(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer, int buffer_s
     }
 }
 
+
+static bool
+function_decide_simple_inlineable(
+    _PyUOpInstruction *func_body_start,
+    _PyUOpInstruction *func_body_end)
+{
+    // Usually means MAKE_CELL or something
+    if (func_body_start->opcode != _RESUME_CHECK) {
+        return false;
+    }
+    func_body_start++;
+    _PyUOpInstruction *curr = func_body_start;
+    while (curr < func_body_end) {
+        int opcode = curr->opcode;
+        // We should be the leaf frame.
+        assert(opcode != _PUSH_FRAME && opcode != _PUSH_FRAME_INLINEABLE);
+        if (opcode == _POP_FRAME) {
+            return true;
+        }
+        if (_PyUop_Flags[curr->opcode] & (HAS_ESCAPES_FLAG | HAS_DEOPT_FLAG | HAS_ERROR_FLAG)) {
+            // Pure overrides error flag.
+            if (!(_PyUop_Flags[curr->opcode] & HAS_PURE_FLAG)) {
+                return false;
+            }
+        }
+        curr++;
+    }
+    Py_UNREACHABLE();
+}
+
+static void
+inline_simple_frames(_PyUOpInstruction *buffer, int buffer_size)
+{
+    bool did_inline = false;
+    for (int pc = 0; pc < buffer_size; pc++) {
+        int opcode = buffer[pc].opcode;
+        switch (opcode) {
+            case _PUSH_FRAME_INLINEABLE: {
+                assert(buffer[pc - 3].opcode == _CHECK_STACK_SPACE);
+                assert(buffer[pc - 2].opcode == _INIT_CALL_PY_EXACT_ARGS);
+                assert(buffer[pc - 1].opcode == _SAVE_RETURN_OFFSET);
+                assert(buffer[pc + 1].opcode == _CHECK_VALIDITY_AND_SET_IP ||
+                    buffer[pc + 1].opcode == _CHECK_VALIDITY);
+                // Skip over the CHECK_VALIDITY when deciding,
+                // as those can be optimized away later.
+                if (!function_decide_simple_inlineable(&buffer[pc + 2], buffer + buffer_size)) {
+                    buffer[pc].opcode = _PUSH_FRAME;
+                    break;
+                }
+                assert(buffer[pc + 2].opcode == _RESUME_CHECK);
+                did_inline = true;
+                uint64_t operand = buffer[pc].operand;
+                int locals_len = (int)(operand >> 32);
+                int stack_len = (int)(operand & 0xFFFFFFFF);
+                REPLACE_OP(&buffer[pc - 3], _GROW_TIER2_FRAME, locals_len + stack_len, 0);
```

Comment:
I have a feeling we can do much better here. We can keep track in the *abstract* frame object how much extra space has already been allocated for inlined functions and only request the difference (if any). Then the GROW instruction would be able to directly do something like this:
```
        op(_GROW_TIER2_FRAME, (--)) {
            DEOPT_IF(tstate->datastack_top + oparg > tstate->datastack_limit);
            tstate->datastack_top += oparg;
        }
```
(Or you can put that logic in `_PyFrame_ConvertToTier2()`.)

Separately, it looks to me like the current calculation of how much is needed should at least deduct the arg count of the call being inlined, because of the frame overlap. (We're still doing the overlap, right?)

Reply:
Yes we're doing the overlap. That's why all the `LOAD_FAST/STORE_FAST` gets updated.

We can't keep track of the frame precisely because we're tracing from backward jumps, not function entry, and also for partial snippets of code. There's no guarantee that this code has not been run before. Because this is partial trace, the code may execute once, side exit, jump back, then execute again, causing the frame to grow unnecessarily big.

Reply:
Hold on. There should **never** be a situation where the depth of the value stack isn't statically known. (Your renumbering of LOAD_FAST etc. relies on this.) When we start we get the stack level passed in (you added that a few months ago).

But I suspect I am not following something crucial (wouldn't be the first time :-). I will check out the code and play with it to help me understand. Sorry.

Reply:
I meant what isn't consistent is how big the extra stack will grow in the future. Consider the following code:

```
_GROW_TIER2_FRAME
xxx
_EXIT_TRACE
```

Lets say outside of the exit trace, we jump backwards and enter the trace again. The same frame will then grow again for a second time, if it doesn't keep track that it has grown once already. This is complicated by the fact that a single frame, can have _multiple_ traces. E.g. I'm not sure how to handle this without keeping track inside the frame:
```
Trace 1:
_GROW_TIER2_FRAME
xxx
_EXIT_TRACE

Re-enter
Trace 2:
_GROW_TIER2_FRAME
xxx
_EXIT_TRACE
```
```
```

Reply:
Okay, now I get it. So what to do about it?

I think that between the address of `frame->localsplus` (which BTW is just a fixed constant from `frame`, though you have floated an idea to change that), `tstate->datastack_top`, and `tstate->datastack_limit`, and assuming the frame is the top frame and `frame->owner == FRAME_OWNED_BY_THREAD`, we have all the information we need to know whether to bump `tstate->datastack_top`.

I believe statically we can always know the offset between `stack_pointer` and `frame->localsplus` (you use this knowledge for patching up `LOAD_FAST` etc.). When the frame is initially created, we allocate space for its `localsplus` array computed from the code object using `code->co_nlocalsplus + code->co_stacksize`. A related value is `code->co_framesize`, which looks to be the size of the frame struct less the `localsplus` array plus the above. In particular, before allocation, the start of the frame is at the old `tstate->datastack_top` and after allocation the latter variable has been updated to point to the end of the frame (including `localsplus`). See `_PyFrame_PushUnchecked` in pycore_frame.h, and the slower fallback, `_PyThreadState_PushFrame` in pystate.c.

I was going to draw some pictures but once I drew them I realized we can just do the following:
```c
        op(_GROW_TIER2_FRAME, (--)) {
            DEOPT_IF(frame->owner != FRAME_OWNED_BY_THREAD);
            DEOPT_IF(stack_pointer + oparg > tstate->datastack_limit);
            assert(tstate->datastack_top < stack_pointer);
            tstate->datastack_top = stack_pointer + oparg;
        }
```
We need to roll back your changes to ceval.c and pycore_frame.h and one or two miscellaneous other thing (e.g. test_sys.py) and it seems to work.

<details>
<summary>(Click here for the pictures I already drew:)</summary>

Initially (immediately after the frame has been allocated) we have:
```
frame      localsplus         datastack_top          datastack_limit
v          v                  v                      v
+----------+--------+---------+----------------------+
| SPECIALS | locals | stack   |     (free space)     |
+----------+--------+---------+----------------------+
                        ^
                        stack_pointer
```

Suppose we are inlining a call to a function that uses `N` locals
and `M` stack items. Drawing this alongside the above, we get:
```
frame      localsplus         datastack_top          datastack_limit
v          v                  v                      v
+----------+--------+---------+----------------------+
| SPECIALS | locals | stack   |     (free space)     |
+----------+--------+---------+----------------------+
                        ^
                        stack_pointer
                        +------+------+
                        |  N   |  M   |
                        +------+------+
```

My proposal is that in the `_GROW_TIER2_FRAME` opcode we effectively
do the following calculation:
```
    if (tstate->datastack_top < stack_pointer + N + M) {
        DEOPT_IF(stack_pointer + N + M > tstate->datastack_limit);
        tstate->datastack_top = stack_pointer + N + M;
    }
```
After this operation we end up with the following situation:
```
frame      localsplus                 datastack_top  datastack_limit
v          v                          v              v
+----------+--------+-----------------+--------------+
| SPECIALS | locals | stack           | (free space) |
+----------+--------+-----------------+--------------+
                        ^             |
                        stack_pointer |
                        +------+------+
                        |  N   |  M   |
                        +------+------+
```

If at some later time we find ourselves in exactly this situation
or in a situation where `stack_pointer + N + M` is less than
`tstate->datastack_top` we will keep the latter unchanged.
```
frame      localsplus                   datastack_top  datastack_limit
v          v                            v              v
+----------+--------+-------------------+--------------+
| SPECIALS | locals | stack             | (free space) |
+----------+--------+-----------------+-+--------------+
                        ^             |
                        stack_pointer |
                        +------+------+
                        |  N   |  M   |
                        +------+------+
```

</details>

Reply:
So my fix didn't quite work, but it almost worked. Brandt found it could be made to work by adding the size of the "unused" portion of the stack to the oparg (when emitting this uop). That surprises me because I had assumed that it worked like this (from my earlier ASCII art):
```
frame      localsplus         datastack_top          datastack_limit
v          v                  v                      v
+----------+--------+---------+----------------------+
| SPECIALS | locals | stack   |     (free space)     |
+----------+--------+---------+----------------------+
                        ^
                        stack_pointer
                        +------+------+
                        |  N   |  M   |
                        +------+------+
```
But apparently it works like this:
```
frame      localsplus         datastack_top          datastack_limit
v          v                  v                      v
+----------+--------+---------+----------------------+
| SPECIALS | locals | stack   |     (free space)     |
+----------+--------+---------+----------------------+
                        ^
                        stack_pointer
                              +------+------+
                              |  N   |  M   |
                              +------+------+
```
Or else there's maybe an off-by-one error related to where `datastack_limit` points?